src/backend.rs

```diff
@@ -0,0 +1,465 @@
+                    loop {
+                        tokio::select! {
+                            result = stream.recv() => {
+                                if result.is_some() {
+                                    tracing::debug!(nsignal, "signal received");
+                                    unsafe {
+                                        // libevent dispatches callbacks with the mapped signal
+                                        libevent_sys::evmap_signal_active_(base.0, nsignal, 1);
+                                    }
+                                } else {
+                                    tracing::error!("signal stream has closed");
+                                    break;
+                                }
+                            },
+                            _ = notify.notified() => break, // terminate the task on notification
+                        }
+                    }
+
+                    // indicate that the task has terminated
+                    tracing::debug!(?nsignal, "I/O task terminated");
+                });
+
+                0
+            }
+            Err(error) => {
+                tracing::error!(?error);
+                -1
+            }
+        }
+    }
+
+    /// Terminates an active signal task
+    #[instrument]
+    fn del_signal(&mut self, nsignal: c_int) -> c_int {
+        tracing::debug!("delete an signal event");
+
+        match self.signal_map.remove(&nsignal) {
+            Some(notify) => {
+                notify.notify_one();
+                0
+            }
+            None => {
+                tracing::warn!("signal not found");
+                -1
+            }
+        }
+    }
+}
+
+/// Wrapper to allow sending of raw event_base pointers to tokio tasks.
+///
+/// This is safe because libevent performs locking internally.
+struct BaseWrapper(pub *mut libevent_sys::event_base);
+
+unsafe impl Send for BaseWrapper {}
+
+/// Injects a tokio backend with the given runtime into the given libevent instance.
+///
+/// The libevent instance will already have an initialized backend. This
+/// exisiting backend is deallocated before being replaced.
+///
+/// A tracing-subscriber may also be initialized if the feature is activated
+/// to enable tracing output when linked to a C program.
+pub fn inject_tokio(mut base: NonNull<libevent_sys::event_base>, runtime: tokio::runtime::Runtime) {
+    #[cfg(feature = "tracing_subscriber")]
+    tracing_subscriber::fmt::init();
+
+    let backend = Box::new(TokioBackend::new(runtime));
+    let base = unsafe { base.as_mut() };
+
+    if let Some(evsel) = unsafe { base.evsel.as_ref() } {
+        if let Some(dealloc) = evsel.dealloc {
+            unsafe {
+                dealloc(base);
+            }
+        }
+    }
+
+    base.evsel = &backend.evsel;
+    base.evsigsel = &backend.evsigsel;
+    base.evbase = Box::into_raw(backend).cast();
+}
+
+/// Convenience function that returns true if the signal event bit is set.
```

Comment:
I would guess the compiler will figure it out but these might be decent candidates for explicitly requesting inlining via `#[inline(always)]`